### PR TITLE
[GEOT-6833] Add missing @Override annotations and set up a QA rule to enforce it

### PR DESCRIPTION
**Pull request by groldan**
https://osgeo-org.atlassian.net/browse/GEOT-6833

- Set up PMD rule
- Add `@Override` annotation to all methods that missed it, using the `-Dall` maven flag

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

**Comment by aaime** [comment]
image

LOL!

**Comment by groldan** [comment]
> image
> 
> LOL!

yeah, I've split it into separate commits for build/*, modules/library|ogc|plugin|extension|unsupported, so at least there are less than 3000 affected files per commit :/

**Comment by aaime** [comment]
@groldan thanks for this cleanup.

There is only one thing that's giving me pause, it's not necessarily wrong mind: "[GEOT-6833] Add missing @override annotations to modules/ogc/*"

These modules are made of generated code (and tendentially, pretty ugly too). [QA is disabled in them](https://github.com/geotools/geotools/blob/main/modules/ogc/pom.xml#L70), because editing them is pointless, then next time we regenerate them the code will be rewritten from scratch.

Long story short... the commit can be kept, and the result will likely stick for years, but if, for any reason, we need to regenerate the models, then it will be gone.

**Comment by aaime** [comment]
Thinking out loud, the PR is so massive that I'm wondering if it would be better to also backport it, to avoid conflicts. Hum... to be honest, I have made a number so far, which cumulatively probably touched the same if not more file, and that did not cause much friction (git be blessed!)... so maybe not... well, just thinking out loud here! :-D

**Comment by aaime** [comment]
@groldan we have a situation that will likely cause lots of conflicts. Back in January [I started working on this PR](https://github.com/geotools/geotools/pull/3318), which required some amount of manual fiddling (to remove the cases where a lambda was less readable and cases where the removal of the anonymous innner class removed PMD suppression or warning suppressions). 

Is your PR mostly done automatically (besides the first and last two commits, it seems like it might be)? If so, could I merge first, and you can re-apply the automated cleanups on top of the new main?

**Comment by groldan** [comment]
@aaime yes, by all means go ahead and merge #3318 first. The hard part of this one are 9a9cd9a3e496633cb761d437c63e1c3f47af6d14 and e3f362246b0990d33fa603932fef76ab0fe05f20, the rest is automated. Thanks.

**Comment by aaime** [comment]
Merged. Lots of conflicts as predicted.

**Comment by groldan** [comment]
> Merged. Lots of conflicts as predicted.

Looks like `git rebase main -X ours` made the trick. e.g.:

**before**:

```java
diff main -- modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
diff --git a/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java b/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
index 4e22adc723..d14dd23b6b 100644
--- a/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
+++ b/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
@@ -32,11 +32,13 @@ public class StagedGraphTraversal extends BasicGraphTraversal {
         m_stage = 0;
     }
 
+    @Override
     public void init() {
         // initialize the nodes of the graph by setting counts to 0
         getGraph()
                 .visitNodes(
                         new GraphVisitor() {
+                            @Override
                             public int visit(Graphable component) {
                                 component.setCount(0);
                                 return (0);
@@ -44,10 +46,12 @@ public class StagedGraphTraversal extends BasicGraphTraversal {
                         });
     }
 
+    @Override
     public boolean isVisited(Graphable g) {
         return (g.getCount() == m_stage);
     }
 
+    @Override
     public void setVisited(Graphable g, boolean visited) {
         g.setCount(visited ? m_stage : -1);
     }
```

> that'd be `diff 770d1c5c e3f36224  -- modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java`

**after**:

```java
diff main -- modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
diff --git a/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java b/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
index cd44876d4d..52351c7cce 100644
--- a/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
+++ b/modules/extension/graph/src/main/java/org/geotools/graph/traverse/basic/StagedGraphTraversal.java
@@ -31,6 +31,7 @@ public class StagedGraphTraversal extends BasicGraphTraversal {
         m_stage = 0;
     }
 
+    @Override
     public void init() {
         // initialize the nodes of the graph by setting counts to 0
         getGraph()
@@ -41,10 +42,12 @@ public class StagedGraphTraversal extends BasicGraphTraversal {
                         });
     }
 
+    @Override
     public boolean isVisited(Graphable g) {
         return (g.getCount() == m_stage);
     }
 
+    @Override
     public void setVisited(Graphable g, boolean visited) {
         g.setCount(visited ? m_stage : -1);
     }
```

**Comment by aaime** [comment]
Nice. Merging!
